### PR TITLE
[SPARK-22418][SQL][TEST] Add test cases for NULL Handling

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

Added a test class to check NULL handling behavior.
The expected behavior is defined as the one of the most well-known databases as specified here: https://sqlite.org/nulls.html.

SparkSQL behaves like other DBs:
 - Adding anything to null gives null -> YES
 - Multiplying null by zero gives null -> YES
 - nulls are distinct in SELECT DISTINCT -> NO
 - nulls are distinct in a UNION -> NO
 - "CASE WHEN null THEN 1 ELSE 0 END" is 0? -> YES
 - "null OR true" is true -> YES
 - "not (null AND false)" is true -> YES
 - null in aggregation are skipped -> YES

## How was this patch tested?

Added test class

### CODE REVIEW

**Comment by mgaido91** [comment]
I think that the failure is related to the lack of changes in code outside tests but I am not sure...

**Comment by hvanhovell** [comment]
retest this please

**Comment by mgaido91** [comment]
I think there is an infra issue. The error for the failed test job seems to me:
```
chmod: cannot access `target/*': No such file or directory
```
I am not sure what is happening...

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
I suspect the problem is related with dropped path in some Jenkins machines (we met this few times before). I roughly guess worker 2, 6 and 7. I reported this to Jenkins admin.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by mgaido91** [comment]
thanks @HyukjinKwon

**Comment by gatorsmile** [comment]
Thank you for taking this JIRA! Could you summarize the behavior of our Spark SQL in the PR description?

File: sql/core/src/test/scala/org/apache/spark/sql/NullHandlingSuite.scala [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.test.SharedSQLContext
+
+case class T1(a: Int, b: Option[Int], c: Option[Int])
+
+/**
+ * This test suite takes https://sqlite.org/nulls.html as a reference.
+ */
+class NullHandlingSuite extends QueryTest with SharedSQLContext {
```
Comment (gatorsmile) [inline sql/core/src/test/scala/org/apache/spark/sql/NullHandlingSuite.scala]: My original proposal is to do it in `SQLQueryTestSuite.scala` like the other `.sql` files. Just curious why you want to write it using Dataset APIs?

Comment (mgaido91) [reply sql/core/src/test/scala/org/apache/spark/sql/NullHandlingSuite.scala]: I wasn't aware of your original proposal, sorry.
I simply prefer it because I think it is less error prone.

I can change it if you want.

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/NullHandlingSuite.scala]: When you create a .sql file, you just need to run the command:
> SPARK_GENERATE_GOLDEN_FILES=1 build/sbt "sql/test-only *SQLQueryTestSuite"

Why you think it is error-prone?

**Comment by gatorsmile** [comment]
> * in SparkSQL, the standard SQL UNION clause is rewritten as UNION + DISTINCT
```Scala
      case SqlBaseParser.UNION if all =>
        Union(left, right)
      case SqlBaseParser.UNION =>
        Distinct(Union(left, right))
```

We follow ANSI SQL if you rewrite it to SQL

File: sql/core/src/test/scala/org/apache/spark/sql/NullHandlingSuite.scala (hunk 1 above)
Comment (mgaido91) [reply sql/core/src/test/scala/org/apache/spark/sql/NullHandlingSuite.scala]: I think that in general writing SQL code is more error prone than writing Scala code because there are no compile checks and I prefer to write Scala code than SQL for this reason. Moreover, I think it is easier to do some things like running for loops over all the possible 0's implementations. With SQL I should have copied and pasted the same query multiple times to achieve the same. But maybe this is not even necessary, we can just check one.

As I said, I wasn't aware that you had a proposal for the location of the checks, thus I just did what I considered the best option. If you want, I can move them to a sql file.

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/NullHandlingSuite.scala]: Maybe using SQL? It will be easier for the others who knew SQL only to understand our NULL handling logics. Thanks again!

File: sql/core/src/test/resources/sql-tests/results/null-handling.sql.out [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+153
+155
+
+
+-- !query 27
+drop table t1
+-- !query 27 schema
+struct<>
+-- !query 27 output
+
```
Comment (gatorsmile) [inline sql/core/src/test/resources/sql-tests/results/null-handling.sql.out]:
```SQL
Table created.

1 row(s) inserted.

1 row(s) inserted.

1 row(s) inserted.

1 row(s) inserted.

1 row(s) inserted.

1 row(s) inserted.

1 row(s) inserted.

Result Set 1
A	B+C
1	0
2	1
3	1
4	2
5	 - 
6	 - 
7	 - 

7 rows selected.

Result Set 2
A+10	B*0
11	0
12	0
13	0
14	0
15	 - 
16	 - 
17	 - 

7 rows selected.

Result Set 3
B
1
 - 
0

3 rows selected.

Result Set 4
B
0
1
 - 

3 rows selected.

Result Set 5
A+20	CASEBWHENCTHEN1ELSE0END
21	1
22	0
23	0
24	1
25	0
26	0
27	0

7 rows selected.

Result Set 6
A+30	CASECWHENBTHEN1ELSE0END
31	1
32	0
33	0
34	1
35	0
36	0
37	0

7 rows selected.

Result Set 7
A+40	CASEWHENB<>0THEN1ELSE0END
41	0
42	0
43	1
44	1
45	0
46	0
47	0

7 rows selected.

Result Set 8
A+50	CASEWHENNOTB<>0THEN1ELSE0END
51	1
52	1
53	0
54	0
55	0
56	0
57	0

7 rows selected.

Result Set 9
A+60	CASEWHENB<>0ANDC<>0THEN1ELSE0END
61	0
62	0
63	0
64	1
65	0
66	0
67	0

7 rows selected.

Result Set 10
A+70	CASEWHENNOT(B<>0ANDC<>0)THEN1ELSE0END
71	1
72	1
73	1
74	0
75	1
76	0
77	0

7 rows selected.

Result Set 11
A+80	CASEWHENB<>0ORC<>0THEN1ELSE0END
81	0
82	1
83	1
84	1
85	0
86	1
87	0

7 rows selected.

Result Set 12
A+90	CASEWHENNOT(B<>0ORC<>0)THEN1ELSE0END
91	1
92	0
93	0
94	0
95	0
96	0
97	0

7 rows selected.

Result Set 13
COUNT(*)	COUNT(B)	SUM(B)	AVG(B)	MIN(B)	MAX(B)
7	4	2	.5	0	1


Result Set 14
A+100
101
102
103
104

4 rows selected.

Result Set 15
A+110
111
112
113
114

4 rows selected.

Result Set 16
A+120
121
122
123
124
126

5 rows selected.

Result Set 17
A+130
132
134

2 rows selected.

Result Set 18
A+140
141
143
145

3 rows selected.

Result Set 19
A+150
151
153
155

3 rows selected.

Table dropped.
```

**Comment by gatorsmile** [comment]
Our results are exactly the same as Oracle. 

LGTM

**Comment by gatorsmile** [comment]
retest this please

**Comment by mgaido91** [comment]
I checked with Postgres and we have the same results of Postgres too.

**Comment by gatorsmile** [comment]
Thanks! Merged to master.
